File: Data Structures & Algorithms/house-robber/submission-2.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        memo = {}

        def dfs(i):
            if i == 0 or i ==1:
                return nums[i]
            if i == -1:
                return 0
            if i in memo:
                return memo[i]

            memo[i] = nums[i] + max(dfs(i-2),dfs(i-3))
            return memo[i]
        
        return max( dfs(len(nums)-1), dfs(len(nums)-2) )
        

File: Data Structures & Algorithms/combination-target-sum/submission-2.cpp
class Solution {
private:
    std::vector<std::vector<int>> res;
    std::vector<int> nums;
    int target;
    
public:
    vector<vector<int>> combinationSum(vector<int>& nums, int target) {

        this->nums = std::move(nums);
        this->target = target;
        std::vector<int> curr_lst;
        dfs(curr_lst,0,0);
        return res;
    }

    void dfs(std::vector<int> &curr_lst,int i, int total){
        if (total == target){
            res.push_back(curr_lst);
            return;
        }
        if (total > target || i >= nums.size()) return;

        curr_lst.push_back(nums[i]);
        dfs(curr_lst,i,total+nums[i]);
        curr_lst.pop_back();

        dfs(curr_lst,i+1,total);
        
        return;
    }
};
